bookwyrm/views/preferences/two_factor_auth.py

```diff
@@ -0,0 +1,175 @@
+            "qrcode": self.create_qr_code(request.user),
+            "form": qr_form,
+        }
+        return TemplateResponse(request, "preferences/2fa.html", data)
+
+    def create_qr_code(self, user):
+        """generate and save a qr code for 2FA"""
+        otp_secret = pyotp.random_base32()
+        # save the secret to the user record - we'll need it to check codes in future
+        user.otp_secret = otp_secret
+        user.save(broadcast=False, update_fields=["otp_secret"])
+        # now we create the qr code
+        provisioning_url = pyotp.totp.TOTP(otp_secret).provisioning_uri(
+            name=user.localname, issuer_name=DOMAIN
+        )
+        qr_code = qrcode.QRCode(image_factory=qrcode.image.svg.SvgPathImage)
+        qr_code.add_data(provisioning_url)
+        qr_code.make(fit=True)
+        img = qr_code.make_image(attrib={"fill": "black"})
+        return img.to_string()
+
+
+@method_decorator(login_required, name="dispatch")
+class Confirm2FA(View):
+    """confirm user's 2FA settings"""
+
+    def post(self, request):
+        """confirm the 2FA works before requiring it"""
+        form = forms.Confirm2FAForm(request.POST, instance=request.user)
+
+        if not form.is_valid():
+            data = {
+                "password_confirmed": True,
+                "qrcode": Edit2FA.create_qr_code(self, request.user),
+                "form": form,
+            }
+            return TemplateResponse(request, "preferences/2fa.html", data)
+
+        # set the user's 2FA setting on
+        request.user.two_factor_auth = True
+        request.user.save(broadcast=False, update_fields=["two_factor_auth"])
+        data = {"form": form, "success": True}
+        return TemplateResponse(request, "preferences/2fa.html", data)
+
+
+@method_decorator(login_required, name="dispatch")
+class Disable2FA(View):
+    """Turn off 2FA on this user account"""
+
+    def get(self, request):
+        """Confirmation page to turn off 2FA"""
+        return TemplateResponse(request, "preferences/disable-2fa.html")
+
+    def post(self, request):
+        """Turn off 2FA on this user account"""
+        request.user.two_factor_auth = False
+        request.user.save(broadcast=False, update_fields=["two_factor_auth"])
+        data = {"form": forms.ConfirmPasswordForm(), "success": True}
+        return TemplateResponse(request, "preferences/2fa.html", data)
+
+
+class LoginWith2FA(View):
+    """Check 2FA code matches before allowing login"""
+
+    def get(self, request):
+        """Display 2FA form"""
+        data = {"form": forms.Confirm2FAForm()}
+        return TemplateResponse(request, "two_factor_auth/two_factor_login.html", data)
+
+    def post(self, request):
+        """Check 2FA code and allow/disallow login"""
+        user = models.User.objects.get(username=request.session["2fa_user"])
+        elapsed_time = datetime.now() - datetime.fromtimestamp(
+            int(request.session["2fa_auth_time"])
```

Comment:
It's probably safer to use `request.session.get("2fa_auth_time")` here (and above) with fallback values, because if a malformed request  is sent (like via cURL) without those keys, there's no need to have a 500 error in the application.